docs/logging.md

```diff
@@ -0,0 +1,88 @@
+# Logging
+
+By default LinuxKit will write onboot and service logs directly to files in
+`/var/log` and `/var/log/onboot`.
+
+If you want the logs to go somewhere else, you must provide an external logging
+service. This is a daemon which runs from `init.d` as part of `init` and which
+listens on the Unix domain socket
+
+```
+/var/run/external-logging.sock
+```
+
+Standard output and standard input file descriptors will be sent (along with
+names) to the logging service. The logging service must read data from these
+fds and process it.
+
+## memlogd: an in-memory circular buffer
+
+`memlogd` is a daemon that keeps logs in a circular buffer in memory. It
+listens on two sockets:
+
+- `/var/run/external-logging.sock`: used for receiving standard out and standard error
+  file descriptors
+- `/var/run/memlogdq.sock`: used for reading the buffered log
+
+When `memlogd` receives a file descriptor it will read from the file
+descriptor and timestamp and append the content to the in-memory log until
+the file is closed.
+
+Since `memlogd` logs to a circular buffer, when the buffer fills then log
+messages will be dropped. Therefore it is recommended that a `service` be
```

Comment:
A good question -- I had a more thorough look at the `memlogd` code. It uses a fixed length `container/ring` initialised here: https://github.com/djs55/linuxkit/blob/cb4817bd419baedd4077bd349d7f989c1942d8d0/pkg/memlogd/cmd/memlogd/main.go#L112 It maintains a pointer into the ring and appends by overwriting and advancing the pointer here: https://github.com/djs55/linuxkit/blob/cb4817bd419baedd4077bd349d7f989c1942d8d0/pkg/memlogd/cmd/memlogd/main.go#L120 so I think it only drops when it cycles round to the beginning and hence drops old messages.

I think there is a memory usage issue. Although the input messages are in a fixed-size ring, there is a per-reader [bytes.Buffer](https://github.com/djs55/linuxkit/blob/cb4817bd419baedd4077bd349d7f989c1942d8d0/pkg/memlogd/cmd/memlogd/main.go#L47) which has no bounds. If the reader is slower than the writer then the data will fill up in memory. Perhaps this should be a ring too?

I think another nice-to-have in the `memlogd` code would be to keep a log message counter, so clients could detect when they've dropped messages (and log "dropped n messages") since it's useful to know when a log message was definitely not written as opposed to maybe written and then dropped.

Reply:
[`bytes.NewBuffer`](https://golang.org/pkg/bytes/#NewBuffer) can be given a `[]byte` with a fixed capacity which should bound it I think.

Reply:
My mistake, that just sets the initial size, it can still grow...

Reply:
bounding the buffers now would be good. Adding a counter later is fine. I'm not sure I understand your design regarding the counter though. I would think the client sending the counter as part of the message and `memlogd` keeping an internal counter allows `memlogd` to detect dropped messages and log it accordingly, ie `memlogd` detects dropped messages instead of the client (as you described above).